contribute/design/1.x/csi/20190805-csi-zfspv-volume-provisioning.md

```diff
@@ -109,48 +109,105 @@ spec:
+provisioner: zfs.csi.openebs.io
+parameters:
+  blocksize: "4k"
+  compression: "on"
+  dedup: "on"
+  thinprovision: "yes"
+  poolname: "pool2"
+```
+In case of same number of volumes on all the nodes for the given pool, it can pick any node and schedule the PV on that.
+
+#### 1.2  Volume Creation
+
+It will create the PV object on scheduled node so that the applcation using that PV always comes to the same node and also it creates the ZFSVolume object for that volume in order to manage the creation of the ZFS dataset. There will be a watcher at each node which will be watching for the ZFSVolume resource which is aimed for them. The watcher is inbuilt into ZFS node-agent. As soon as ZFSVolume object is created for a node, the corresponding watcher will get the add event and it will create the ZFS dataset with all the volume properties from ZFSVolume custom resource. It will get the pool name from the ZFSVolume custom resource and creates the volume in that pool with pvc name.
 
 ### 2. CSI Node Publish
-When kubernetes schedules the application pod according to the Topology mentioned in the StorageClass, The CSI will get a NodePublish event. It will get all the volume properties from ZfsVolume custom resource and it will fire zfs create command and creates the volume. It will get the pool name from the ZfsVolume custom resource and create a volume on that pool with pvc name. Then it will format and mount the created zvol to the desired location and return.
+When the application pod is scheduled on a node, the CSI node-agent will get a NodePublish event. The driver will get the device path from the ZFSVolume custome resource and try to mount the file system as per user request given in the storage class. Once the ZFS volume dataset is created it will put a finalizer and return successful for the NodePublish event.
 
 - the kubernetes managed ZFS volume will look like this
 
 ```yaml
-apiVersion: v1
-items:
-- apiVersion: openebs.io/v1alpha1
-  kind: ZFSVolume
-  metadata:
-    creationTimestamp: "2019-08-25T08:02:14Z"
-    generation: 2
-    name: pvc-a6855135-c70e-11e9-8fa2-42010a80012d
-    namespace: openebs
-    resourceVersion: "13735"
-    selfLink: /apis/openebs.io/v1alpha1/namespaces/openebs/zfsvolumes/pvc-a6855135-c70e-11e9-8fa2-42010a80012d
-    uid: a691916c-c70e-11e9-8fa2-42010a80012d
-  spec:
-    blocksize: 4k
-    capacity: "4294967296"
-    compression: "on"
-    dedup: "on"
-    devicePath: /dev/zvol/zfspv-pool/pvc-a6855135-c70e-11e9-8fa2-42010a80012d
-    ownerNodeID: gke-user-zfspv-default-pool-354050c7-wl8v
-    poolName: zfspv-pool
-    thinprovison: "yes"
-    volName: pvc-a6855135-c70e-11e9-8fa2-42010a80012d
-kind: List
+apiVersion: openebs.io/v1alpha1
+kind: ZFSVolume
 metadata:
-  resourceVersion: ""
-  selfLink: ""
+  creationTimestamp: "2019-11-03T06:08:02Z"
+  finalizers:
+  - zfs.openebs.io/finalizer
+  generation: 1
+  labels:
+    kubernetes.io/nodename: gke-pawan-zfspv-default-pool-8b577544-5wgl
```

Comment:
what is the significance of this?

Reply:
it will be same as OwnerNodeID, it is there so that we can query for a nodename to get all the volumes provisioned on it.